### PR TITLE
Adding MPTrie Proof API to sdk

### PR DESCRIPTION
**Pull request by gennadylaventman**
Signed-off-by: Gennady Laventman <gennady@il.ibm.com>

### CODE REVIEW

File: pkg/bcdb/ledger_test.go [hunk 1]
```diff
@@ -267,3 +269,117 @@ func TestGetTransactionReceipt(t *testing.T) {
 		})
 	}
 }
+
+func TestGetStateProof(t *testing.T) {
```
Comment (tock-ibm) [inline pkg/bcdb/ledger_test.go:273]: It would be nice and instructive to add a test that completes the circle from a TX, TxReceipt, and then data proof, to connect the dots...

File: pkg/bcdb/ledger_test.go [hunk 2]
```diff
@@ -267,3 +269,117 @@ func TestGetTransactionReceipt(t *testing.T) {
...
+	for i := 0; i < 10; i++ {
+		keys := make([]string, 0)
+		values := make([]string, 0)
+		for j := 0; j < 10; j++ {
+			keys = append(keys, fmt.Sprintf("key%d_%d", i, j))
+			values = append(values, fmt.Sprintf("value%d_%d", i, j))
+		}
+		blockTx := putMultipleKeysAndValues(t, keys, values, "alice", aliceSession)
+		txEnvelopesPerBlock = append(txEnvelopesPerBlock, blockTx)
+	}
```
Comment (tock-ibm) [inline pkg/bcdb/ledger_test.go:292]: This has a chance of being a flaky test because there might be less than 10 TXs in a block even though 10 is permitted...

File: pkg/bcdb/ledger.go [hunk 3]
```diff
@@ -96,3 +97,26 @@ func (l *ledger) GetTransactionReceipt(txId string) (*types.TxReceipt, error) {
 
 	return resEnv.GetResponse().GetReceipt(), nil
 }
+
+func (l *ledger) GetDataProof(blockNum uint64, dbName, key string, isDeleted bool) (*state.Proof, error) {
```
Comment (tock-ibm) [inline pkg/bcdb/ledger.go:101]: I am not entirely sure `blockNum` is better than `Version` for the API, even though internally we only use the blockNum and not the txIndex...

Comment (gennadylaventman) [reply pkg/bcdb/ledger.go:101]: For me it is confusing - if I provide version with tx index before updating tx, I will expect old value, not new one...

File: pkg/bcdb/ledger_test.go (hunk 1 above)
Comment (gennadylaventman) [reply pkg/bcdb/ledger_test.go:273]: Will take care of that

File: pkg/bcdb/ledger_test.go (hunk 2 above)
Comment (gennadylaventman) [reply pkg/bcdb/ledger_test.go:292]: You right, I will look into test to update it

File: pkg/bcdb/ledger_test.go (hunk 1 above)
Comment (gennadylaventman) [reply pkg/bcdb/ledger_test.go:272]: see `examples/api/state_proof`

File: pkg/bcdb/ledger.go (hunk 3 above)
Comment (tock-ibm) [reply pkg/bcdb/ledger.go:101]: ok

File: examples/api/state_proof/state_proof.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	fmt.Println("Adding key, value: key2, val2 to the database")
+	err = tx.Put("bdb", "key2", []byte("val2"), nil)
+	if err != nil {
+		fmt.Printf("Adding new key to database failed, reason: %s\n", err.Error())
+		return
+	}
+
+	fmt.Println("Committing transaction")
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:74]: fmt.Println("Committing transaction synchronously")

File: examples/api/state_proof/state_proof.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	txID, receipt1, err := tx.Commit(true)
+	if err != nil {
+		fmt.Printf("Commit failed, reason: %s\n", err.Error())
+		return
+	}
+	fmt.Printf("Transaction number %s committed successfully in block %d\n", txID, receipt1.Header.BaseHeader.Number)
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:80]: Transaction ID %s

File: examples/api/state_proof/state_proof.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+		return
+	}
+
+	fmt.Println("Committing transaction")
+	txID, receipt2, err := tx.Commit(true)
+	if err != nil {
+		fmt.Printf("Commit failed, reason: %s\n", err.Error())
+		return
+	}
+	fmt.Printf("Transaction number %s committed successfully in block number %d\n", txID, receipt2.Header.BaseHeader.Number)
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:110]: Transaction ID %s

File: examples/api/state_proof/state_proof.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	key := "key1"
+	value := "val1"
+	fmt.Printf("Verify value in db %s, key %s in block %d\n", dbName, key, receipt1.Header.BaseHeader.Number)
+	stateProof, err := ledger.GetDataProof(receipt1.Header.BaseHeader.Number, "bdb", "key1", false)
+	if err != nil {
+		fmt.Printf("Can't get state proof ledger, reason: %s\n", err.Error())
+		return
+	}
+
+	fmt.Printf("Calculate leaf node hash for db %s, key %s, value [%s] in state trie\n", dbName, key, value)
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:128]: Calculate expected leaf node hash

File: examples/api/state_proof/state_proof.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+	leafHash, err := bcdb.CalculateLeafHash(dbName, key, []byte(value))
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:129]: expectedLeafHash

File: examples/api/state_proof/state_proof.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s] is part of state trie at block %d\n", value, receipt1.Header.BaseHeader.Number)
+	} else {
+		fmt.Printf("Value [%s] is not part of state trie at block %d\n", value, receipt1.Header.BaseHeader.Number)
+	}
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:145]: insert a comment that describes what was done and what is expected...

Comment (tock-ibm) [reply examples/api/state_proof/state_proof.go:145]: I.e., 
- value "val1" is part of the state trie at block K, lets prove that... 
- value "val0" is part of the state trie at block L, lets prove that... 
- value "val0" is NOT part of the state trie at block M, lets prove that..
-etc

otherwise it is pretty hard to follow.

File: pkg/bcdb/data_tx_context_test.go [hunk 10]
```diff
@@ -694,7 +694,7 @@ func createDB(t *testing.T, dbName string, session DBSession) {
 	require.True(t, exist)
 }
 
-func putKeySync(t *testing.T, dbName, key string, value string, user string, session DBSession) {
+func putKeySync(t *testing.T, dbName, key string, value string, user string, session DBSession) *types.TxReceipt{
```
Comment (tock-ibm) [inline pkg/bcdb/data_tx_context_test.go:697]: gofmt

File: examples/api/state_proof/state_proof.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+	fmt.Printf("Calculate leaf node hash for db %s, key %s, value [%s] in state trie\n", dbName, key, value)
+	leafHash, err = bcdb.CalculateLeafHash(dbName, key, []byte(value))
+	if err != nil {
+		fmt.Printf("Can't calculate leaf node hash in state trie, reason: %s\n", err.Error())
+		return
+	}
+
+	isCorrect, err = stateProof.Verify(leafHash, receipt2.Header.StateMerkelTreeRootHash, false)
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:181]: what about an example on a deleted value?

File: pkg/bcdb/db.go [hunk 12]
```diff
@@ -61,6 +62,8 @@ type Ledger interface {
 	GetTransactionProof(blockNum uint64, txIndex int) (*TxProof, error)
 	// GetTransactionReceipt return block header where tx is stored and tx index inside block
 	GetTransactionReceipt(txId string) (*types.TxReceipt, error)
+	// GetDataProof returns path in Merkle-Patricia Trie from leaf (value) to trie root stored in BlockHeader
+	GetDataProof(blockNum uint64, dbName, key string, isDeleted bool) (*state.Proof, error)
```
Comment (tock-ibm) [inline pkg/bcdb/db.go:66]: explain more how to use it, reference that calculate and verify methods

File: pkg/bcdb/ledger.go [hunk 13]
```diff
@@ -96,3 +97,38 @@ func (l *ledger) GetTransactionReceipt(txId string) (*types.TxReceipt, error) {
...
+		l.logger.Errorf("failed to execute state proof query %s, due to %s", path, err)
+		return nil, err
+	}
+
+	// TODO: signature verification
+
+	return state.NewProof(resEnv.GetResponse().GetPath()), nil
+}
+
+func CalculateLeafHash(dbName, key string, value []byte) ([]byte, error) {
```
Comment (tock-ibm) [inline pkg/bcdb/ledger.go:124]: Document the interface, or here...

File: examples/api/state_proof/state_proof.go (hunk 9 above)
Comment (gennadylaventman) [reply examples/api/state_proof/state_proof.go]: Added comment about what we want to do before querying server for proof

File: pkg/bcdb/data_tx_context_test.go (hunk 10 above)
Comment (gennadylaventman) [reply pkg/bcdb/data_tx_context_test.go]: done

File: examples/api/state_proof/state_proof.go (hunk 11 above)
Comment (gennadylaventman) [reply examples/api/state_proof/state_proof.go]: Added deleted value example

File: examples/api/state_proof/state_proof.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+		return err
+	}
+
+	ledger, err := session.Ledger()
+	if err != nil {
+		fmt.Printf("Can't get access to ledger, reason: %s\n", err.Error())
+		return err
+	}
+
+	// key1 has value val1 in first data block, lets prove it
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:42]: key1 has value val1 after the first transaction (receipt1), lets prove it

File: examples/api/state_proof/state_proof.go [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+		return err
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt1.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		fmt.Printf("Value [%s, %s, %s] is not part of state trie at block %d\n", dbName, key, value, receipt1.GetHeader().GetBaseHeader().GetNumber())
+	}
+
+	// key1 has value val0 in first data block, lets prove it
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:77]: // key1 has value val0 after the second transaction (receipt2), lets prove it

File: examples/api/state_proof/state_proof.go [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+	// Value in second block
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:78]: // Value to verify:

File: examples/api/state_proof/state_proof.go [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+	value = "val0"
+	// Lets query server for proof for value associated with key key1 in second data block
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:80]: ... in the data block pointed to by receipt2

File: examples/api/state_proof/state_proof.go [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	// key1 has value val1 in first data block, lets prove it
+	// Value to verify
+	dbName := "bdb"
+	key := "key1"
+	value := "val1"
+	deleted := false
+
+	// Lets query server for proof for value associated with key key1 in first data block
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:49]: ... in the data block pointed to by receipt1

File: examples/api/state_proof/state_proof.go [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return err
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		fmt.Printf("Value [%s, %s, %s] is not part of state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	}
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:106]: you expect correct, right? so exit with an error it it is not

File: examples/api/state_proof/state_proof.go [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return err
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt2.Header.BaseHeader.Number)
+	} else {
+		fmt.Printf("Value [%s, %s, %s] is not part of state trie at block %d\n", dbName, key, value, receipt2.Header.BaseHeader.Number)
+	}
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:129]: You expect incorrect, right? so exit with an error if it is correct

File: examples/api/state_proof/state_proof.go [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+
+	// Key key2 value val2 was deleted during second block, lets prove it
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:131]: deleted in the second transaction

File: examples/api/state_proof/state_proof.go [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return nil
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] was deleted in state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		fmt.Printf("Value [%s, %s, %s] wasn't deleted in state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	}
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:161]: you expect correct, right? so exit with an error if it isn't

File: examples/api/state_proof/state_proof.go [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+	"github.com/hyperledger-labs/orion-sdk-go/pkg/config"
+	"github.com/hyperledger-labs/orion-server/pkg/logger"
+	"github.com/hyperledger-labs/orion-server/pkg/types"
+)
+
+func main() {
+	ExcecuteStateTrieExample()
+}
+
+func ExcecuteStateTrieExample() error {
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:17]: that is good, this way we can use it in testing. just make sure to return with error when things don't go as expected within.

File: examples/api/state_proof/state_proof.go [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+	isCorrect, err := stateProof.Verify(valueHash, receipt1.GetHeader().GetStateMerkelTreeRootHash(), deleted)
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return err
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt1.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		errors.New(fmt.Sprintf("Value [%s, %s, %s] is not part of state trie at block %d\n", dbName, key, value, receipt1.GetHeader().GetBaseHeader().GetNumber()))
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:76]: return errors.Errorf(...)

File: examples/api/state_proof/state_proof.go [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+	stateProof, err = ledger.GetDataProof(receipt2.GetHeader().GetBaseHeader().GetNumber(), dbName, key, deleted)
+	if err != nil {
+		fmt.Printf("Can't get state proof ledger, reason: %s\n", err.Error())
+		return err
+	}
+
+	fmt.Printf("Calculate value hash in Merkle-Patricia trie (stored as field in Branch or Value nodes) for db %s, key %s, value %s in state trie, by concatinating db, key and value hashes\n", dbName, key, value)
+	valueHash, err = bcdb.CalculateValueHash(dbName, key, []byte(value))
+	if err != nil {
+		fmt.Printf("Calculate value hash in Merkle-Patricia trie , reason: %s\n", err.Error())
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:93]: Failed to calculate...

File: examples/api/state_proof/state_proof.go [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+	isCorrect, err = stateProof.Verify(valueHash, receipt2.GetHeader().GetStateMerkelTreeRootHash(), deleted)
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return err
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		return errors.New(fmt.Sprintf("Value [%s, %s, %s] is not part of state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber()))
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:107]: return errors.Errorf(...)

File: examples/api/state_proof/state_proof.go [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+
+	fmt.Println("Verify proof using value hash and Merkle-Patricia trie root stored in block header")
+	isCorrect, err = stateProof.Verify(valueHash, receipt2.GetHeader().GetStateMerkelTreeRootHash(), deleted)
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return err
+	}
+
+	if isCorrect {
+		return errors.New(fmt.Sprintf("Value [%s, %s, %s] is part of state trie at block %d\n", dbName, key, value, receipt2.Header.BaseHeader.Number))
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:128]: return errors.Errorf(...)

File: examples/api/state_proof/state_proof.go [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+	isCorrect, err = stateProof.Verify(valueHash, receipt2.GetHeader().GetStateMerkelTreeRootHash(), deleted)
+	if err != nil {
+		fmt.Printf("Can't verify value in state trie, reason: %s\n", err.Error())
+		return nil
+	}
+
+	if isCorrect {
+		fmt.Printf("Value [%s, %s, %s] was deleted in state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber())
+	} else {
+		errors.New(fmt.Sprintf("Value [%s, %s, %s] wasn't deleted in state trie at block %d\n", dbName, key, value, receipt2.GetHeader().GetBaseHeader().GetNumber()))
```
Comment (tock-ibm) [inline examples/api/state_proof/state_proof.go:162]: return errors.Errorf(...)
